module/zfs/arc.c

```diff
@@ -4225,15 +4225,17 @@ static uint64_t
 arc_evict_adj(uint64_t frac, uint64_t total, uint64_t up, uint64_t down,
     uint_t balance)
 {
-	if (total < 8 || up + down == 0)
+	if (total < 32 || up + down == 0)
 		return (frac);
 
 	/*
-	 * We should not have more ghost hits than ghost size, but they
-	 * may get close.  Restrict maximum adjustment in that case.
+	 * We should not have more ghost hits than ghost size, but they may
+	 * get close.  To avoid overflows below up/down should not be bigger
+	 * than 1/5 of total.  But to limit maximum adjustment speed restrict
+	 * it some more.
 	 */
-	if (up + down >= total / 4) {
-		uint64_t scale = (up + down) / (total / 8);
+	if (up + down >= total / 16) {
+		uint64_t scale = (up + down) / (total / 32);
```

Comment:
IMHO it's quite unclear why these ratios are guaranteed to avoid overflow while the previous ones weren't.  Or how they related to the stated 1/5 (or indeed where the magic 1/5 comes from).

Reply:
The 5 comes from `frac >= 4 * ofrac` below (1 + 4).  There was a window between 1/5 and 1/4 when both `up + down >= total / 4` and `frac >= 4 * ofrac` were false and so neither scaled `up` down, allowing `frac + up` to overflow over `1ULL << 32`.